File: 20_engine_editor_00_core_and_application_layer/src/Lua/audio/studio/audio_studio_event.lua
require("lua_extension")
require("cpp_class")

AudioStudioEvent=class("AudioStudioEvent",CppClass)

function AudioStudioEvent:ctor(cpp_class_instance)
    AudioStudioEvent.super.ctor(self,cpp_class_instance)
end

--- 实例化C++ Class
function AudioStudioEvent:InitCppClass()
    self.cpp_class_instance_=Cpp.AudioStudioEvent()
end

--- 设置Event参数
--- @param name string @参数名
--- @param value number
--- @param ignore_seek_speed boolean
function AudioStudioEvent:SetParameterByName(name,value,ignore_seek_speed)
    return self.cpp_class_instance_:SetParameterByName(name,value,ignore_seek_speed)
end

--- 设置Event 3D属性
--- @param x number
--- @param y number
--- @param z number
function AudioStudioEvent:Set3DAttribute(x,y,z)
    return self.cpp_class_instance_:Set3DAttribute(x,y,z)
end

--- 播放Event实例
function AudioStudioEvent:Start()
    return self.cpp_class_instance_:Start()
end

function AudioStudioEvent:Stop()
    return self.cpp_class_instance_:Stop()
end

--- 暂停
function AudioStudioEvent:Pause()
    return self.cpp_class_instance_:Pause()
end

File: 18_multithread_render_03_engine_render_queue/src/Engine/Audio/Studio/AudioStudioEvent.cpp
#include "Engine/Audio/Studio/AudioStudioEvent.hpp"
#include "Engine/Component/GameObject.hpp"
#include "Engine/Component/Transform.hpp"
#include "Engine/Utils/Debug.hpp"

AudioStudioEvent::AudioStudioEvent() {}
AudioStudioEvent::~AudioStudioEvent(){}


FMOD_RESULT AudioStudioEvent::SetParameterByName(const char *name, float value, bool ignore_seek_speed) {
    return FMOD_Studio_EventInstance_SetParameterByName(event_instance_,name,value,ignore_seek_speed);
}

FMOD_RESULT AudioStudioEvent::Set3DAttribute(float x,float y,float z){
    FMOD_3D_ATTRIBUTES attributes = { { x,y,z } };
    return FMOD_Studio_EventInstance_Set3DAttributes(event_instance_,&attributes);
}

void AudioStudioEvent::Start() {
    if(event_instance_== nullptr){
        DEBUG_LOG_ERROR("event_instance_== nullptr");
        return;
    }
    FMOD_RESULT result=FMOD_Studio_EventInstance_Start(event_instance_);
    if(result!=FMOD_OK){
        DEBUG_LOG_ERROR("FMOD_Studio_EventInstance_Start result:{}",int(result));
        return;
    }
}

void AudioStudioEvent::Pause() {
    if(event_instance_== nullptr){
        DEBUG_LOG_ERROR("event_instance_== nullptr");
        return;
    }
    FMOD_RESULT result=FMOD_Studio_EventInstance_SetPaused(event_instance_,true);
    if(result!=FMOD_OK){
        DEBUG_LOG_ERROR("FMOD_Studio_EventInstance_SetPaused result:{}",int(result));
        return;
    }
}

void AudioStudioEvent::Stop() {
    if(event_instance_== nullptr){
        DEBUG_LOG_ERROR("event_instance_== nullptr");
        return;
    }
    FMOD_RESULT result=FMOD_Studio_EventInstance_Stop(event_instance_,FMOD_STUDIO_STOP_MODE::FMOD_STUDIO_STOP_IMMEDIATE);
    if(result!=FMOD_OK){
        DEBUG_LOG_ERROR("FMOD_Studio_EventInstance_Stop result:{}",int(result));
        return;
    }
}
